src/cost_estimates/pessimistic.rs

```diff
@@ -0,0 +1,304 @@
+    estimate_key TEXT PRIMARY KEY,
+    current_value NUMBER NOT NULL,
+    samples TEXT NOT NULL
+)";
+
+iterable_enum!(CostField {
+    RuntimeCost,
+    WriteLength,
+    WriteCount,
+    ReadLength,
+    ReadCount,
+});
+
+impl CostField {
+    /// Select `self` out of the given ExecutionCost
+    fn select_key(&self, from_cost: &ExecutionCost) -> u64 {
+        match self {
+            CostField::RuntimeCost => from_cost.runtime,
+            CostField::WriteLength => from_cost.write_length,
+            CostField::WriteCount => from_cost.write_count,
+            CostField::ReadLength => from_cost.read_length,
+            CostField::ReadCount => from_cost.read_count,
+        }
+    }
+}
+
+impl std::fmt::Display for CostField {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CostField::RuntimeCost => write!(f, "runtime"),
+            CostField::WriteLength => write!(f, "write-length"),
+            CostField::WriteCount => write!(f, "write-count"),
+            CostField::ReadLength => write!(f, "read-length"),
+            CostField::ReadCount => write!(f, "read-count"),
+        }
+    }
+}
+
+impl FromSql for Samples {
+    fn column_result(
+        sql_value: rusqlite::types::ValueRef<'_>,
+    ) -> rusqlite::types::FromSqlResult<Self> {
+        let json_value = JsonValue::column_result(sql_value)?;
+        let items = serde_json::from_value(json_value).map_err(|_e| {
+            error!("Failed to parse PessimisticEstimator sample from SQL");
+            FromSqlError::InvalidType
+        })?;
+        Ok(Samples { items })
+    }
+}
+
+impl Samples {
+    fn to_json(&self) -> JsonValue {
+        JsonValue::from(self.items.as_slice())
+    }
+
+    /// Add a new sample to this struct. The pessimistic sampler only adds to the sample set
+    ///  if the sample set is less than SAMPLE_SIZE or the new sample is greater than the current min.
+    /// Boolean return indicates whether or not the sample was included.
+    fn update_with(&mut self, sample: u64) -> bool {
+        if self.items.len() < SAMPLE_SIZE {
+            self.items.push(sample);
+            return true;
+        }
+
+        let (min_index, min_val) = match self
+            .items
+            .iter()
+            .enumerate()
+            .min_by_key(|(_i, value)| *value)
+        {
+            None => {
+                unreachable!("Should find minimum if len() >= SAMPLE_SIZE");
+            }
+            Some(x) => x,
+        };
+
+        if sample > *min_val {
+            self.items[min_index] = sample;
+            return true;
+        }
+
+        return false;
+    }
+
+    /// Return the integer mean of the sample, uses iterative
+    /// algorithm to avoid overflow. The iterative algorithm
+    /// does have some error around *underflows* on the update,
+    /// but only when the new value is close to the average relative
+    /// to the window size.
+    fn mean(&self) -> u64 {
```

Comment:
I'm not positive it would overflow u64 current, but it's possible if the block limits changed, the resolution of the scalar changed, or the window size changed.

Computing as a float is certainly possible, though.